### PR TITLE
ci: Fail PRs if branch VERSION.txt is not open/dev

### PR DESCRIPTION
**Pull request by phlax**
Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
this requires a bit of testing

File: .azure-pipelines/pipelines.yml [hunk 1]
```diff
@@ -113,10 +113,23 @@ stages:
       name: changed
 
     - bash: |
+        set -x
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:116]:
```suggestion
        set -e
```

File: .azure-pipelines/pipelines.yml [hunk 2]
```diff
@@ -113,10 +113,23 @@ stages:
...
         VERSION_DEV="$(cat VERSION.txt | cut -d- -f2)"
         if [[ $VERSION_DEV == "dev" ]]; then
             echo "##vso[task.setvariable variable=isDev;isoutput=true]true"
         else
+            if [[ $(Build.Reason) != "PullRequest" ]]; then
+                # Check to make sure that it was this PR that changed the version, otherwise fail
+                # as the branch needs to be reopened first.
+                # NB: this will not stop a PR that has already passed checks from being landed.
+                DIFF_TARGET_BRANCH="origin/$(System.PullRequest.TargetBranch)"
+                DIFF_REF="$(git merge-base HEAD "${DIFF_TARGET_BRANCH}")"
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:126]:
```suggestion
                DIFF_REF="$(git merge-base HEAD "${DIFF_TARGET_BRANCH}" VERSION.txt)"
```

File: .azure-pipelines/pipelines.yml [hunk 3]
```diff
@@ -113,10 +113,23 @@ stages:
       name: changed
 
     - bash: |
+        set -x
         VERSION_DEV="$(cat VERSION.txt | cut -d- -f2)"
         if [[ $VERSION_DEV == "dev" ]]; then
             echo "##vso[task.setvariable variable=isDev;isoutput=true]true"
         else
+            if [[ $(Build.Reason) != "PullRequest" ]]; then
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:121]:
```suggestion
            if [[ $(Build.Reason) == "PullRequest" ]]; then
```

**Comment by phlax** [comment]
tested in #25851 and appears to work as expected (now)

**Review by zuercher — APPROVED** [review]

**Pull request merged**